syntax = "proto2";
package protobuf;
import "PlayerState.proto";
import "WorldAttribute.proto";
import "PlayerSummary.proto";
import "Event.proto";
import "UdpConfig.proto";
import "TcpConfig.proto";
enum IPProtocol {
	UDP = 1;
	TCP = 2;
}
enum ExpungeReason {
	NOT_EXPUNGED = 0;
	WORLD_FULL = 1;
	ROADS_CLOSED = 2;
}
message ServerToClient {
    optional int64 server_realm = 1;
    optional int64 player_id = 2;
    optional int64 world_time = 3;
    optional int32 seqno = 4;
    optional int32 stc_f5 = 5; //low-priority world time sync algo (not investigated yet, maybe deprecated) in WorldClockService::calculateOneLegLatency
    // 6,7: absent
    repeated PlayerState states = 8;
    repeated WorldAttribute updates = 9;
    repeated int64 stc_f10 = 10;
    optional bool stc_f11 = 11; //=true???
    optional string zc_local_ip = 12;
    optional int64 stc_f13 = 13;
    optional int32 zwifters = 14;
    optional int32 zc_local_port = 15;
    optional IPProtocol zc_protocol = 16;
    optional int64 cts_latency = 17; //high-priority world time sync algo in WorldClockService::calculateOneLegLatency
    optional int32 num_msgs = 18;
    optional int32 msgnum = 19;
    optional bool has_simult_login = 20; //UdpClient::disconnectionRequested due to simultaneous login (1); OR simultaneous login ceased (0)
    optional PlayerSummaries player_summaries = 21; //tag426
    // 22 absent
    optional EventSubgroupPlacements ev_subgroup_ps = 23; //tag442
    optional UdpConfig udp_config = 24; //tag450
    optional UdpConfigVOD udp_config_vod_1 = 25; //tag458
    optional ExpungeReason expunge_reason = 26; //tag464 UdpClient::receivedExpungeReason
    optional UdpConfigVOD udp_config_vod_2 = 27; //tag474
    repeated PlayerState player_states = 28; //tag482
    optional TcpConfig tcp_config = 29; //tag490
    repeated int64 ack_subs_segm = 30; //tag496 TcpClient::processSubscribedSegment
    optional uint32 stc_f31 = 31; //tag508
    optional bytes zc_key = 32; //tag642
}
